lib/client.js

```diff
@@ -1278,6 +1281,128 @@ function _resume (client, sync) {
+  socket[kWriting] = true
+
+  let callback = noop
+  function onDrain (err) {
+    if (!error && err) {
+      error = err
+    }
+    const actualCallback = callback
+    callback = noop
+    actualCallback()
+  }
+
+  const waitForDrain = () => new Promise((resolve) => {
+    assert(callback === noop)
+    callback = resolve
+  })
+
+  let bytesWritten = 0
+  let error
+  try {
+    socket.on('error', onDrain)
+      .on('drain', onDrain)
+    for await (const chunk of body) {
+      if (error || socket.destroyed) {
+        return
+      }
+
+      const len = Buffer.byteLength(chunk)
+      if (!len) {
+        continue
+      }
+
+      // TODO: What if not ended and bytesWritten === contentLength?
+      // We should defer writing chunks.
+      if (contentLength !== null && bytesWritten + len > contentLength) {
+        if (client[kStrictContentLength]) {
+          throw new RequestContentLengthMismatchError()
+        }
+
+        process.emitWarning(new RequestContentLengthMismatchError())
+      }
+
+      if (bytesWritten === 0) {
+        if (!expectsPayload) {
+          socket[kReset] = true
+        }
+
+        if (contentLength === null) {
+          socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+        } else {
+          socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+        }
+      }
+
+      if (contentLength === null) {
+        socket.write(`\r\n${len.toString(16)}\r\n`, 'ascii')
+      }
+
+      bytesWritten += len
+
+      if (socket.write(chunk)) {
+        continue
+      } else if (socket.destroyed || error) {
+        return
+      } else {
+        await waitForDrain()
+        if (error) {
+          return
+        }
+      }
+    }
+  } catch (err) {
+    error = err
+  } finally {
+    socket[kWriting] = false
+
+    if (!error && contentLength !== null && bytesWritten !== contentLength) {
+      if (client[kStrictContentLength]) {
+        error = new RequestContentLengthMismatchError()
+      } else {
+        process.emitWarning(new RequestContentLengthMismatchError())
+      }
+    }
+
+    socket.off('error', onDrain)
+      .off('drain', onDrain)
+    if (error) {
+      assert(client[kRunning] <= 1, 'pipeline should only contain this request')
+      util.destroy(socket, error)
+    }
+
+    if (!socket.destroyed) {
+      if (bytesWritten === 0) {
+        if (expectsPayload) {
+          // https://tools.ietf.org/html/rfc7230#section-3.3.2
+          // A user agent SHOULD send a Content-Length in a request message when
+          // no Transfer-Encoding is sent and the request method defines a meaning
+          // for an enclosed payload body.
+
+          socket.write(`${header}content-length: 0\r\n\r\n\r\n`, 'ascii')
+        } else {
+          socket.write(`${header}\r\n`, 'ascii')
+        }
+      } else if (contentLength === null) {
+        socket.write('\r\n0\r\n\r\n', 'ascii')
+      }
```

Comment:
This should only happen if socket didn't error so you can move it inside the try clause.